main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -2486,18 +2488,28 @@ object Resolver {
   /**
     * Returns the method reflection object for the given `className`, `methodName`, and `signature`.
     */
-  private def lookupJvmMethod(className: String, methodName: String, signature: List[Type], static: Boolean, loc: SourceLocation)(implicit flix: Flix): Validation[Method, ResolutionError] = {
+  private def lookupJvmMethod(className: String, methodName: String, signature: List[Type], retTpe: Type, static: Boolean, loc: SourceLocation)(implicit flix: Flix): Validation[Method, ResolutionError] = {
     // Lookup the class and signature.
     flatMapN(lookupJvmClass(className, loc), lookupSignature(signature, loc)) {
       case (clazz, sig) => try {
         // Lookup the method with the appropriate signature.
         val method = clazz.getDeclaredMethod(methodName, sig: _*)
 
         // Check if the method should be and is static.
-        if (static == Modifier.isStatic(method.getModifiers))
-          method.toSuccess
-        else
+        if (static != Modifier.isStatic(method.getModifiers))
           throw new NoSuchMethodException()
+        else 
+          // Check that the return type of the method matches the declared type.
```

Comment:
Overall approach looks: 👍 

A few things we should resolve:

- We have to call Type.eraseAliases on retTpe (I think).

- In Java, a method is uniquely determined by its name and formal parameter types, but not its return type. Hence we should have an new error message "MismatchedReturnType" that says we found the method, but the declared return type does not match the actual return type.

- We should do an explicit pattern match on the pair `((Type.getFlixType(method.getReturnType), Type.erase..(retTpe)` and handle these cases: `(Int8, Int8) => // OK` and so forth for all of these types: Bool, Char, FloatX, IntX, BigInt, String, and Native. Everything else goes through unchecked.

Reply:
Regarding "We have to call Type.eraseAliases on retTpe (I think)." what is the test that I should write to confirm/deny that this is the right thing to do?

Reply:
> Regarding "We have to call Type.eraseAliases on retTpe (I think)." what is the test that I should write to confirm/deny that this is the right thing to do?

You can define a type alias and use it as the return type. If Magnus is right, then you'll get an error. For example,

```
type alias MyString = ##java.lang.String
...
import new ##java.lang.String(): MyString \ Pure as newString; // This is valid because MyString = String but might not work if we don't erase aliases
```

Reply:
That makes sense. Thanks 👍

Reply:
If you make the above changes then I think we can merge this 👍

Reply:
So I've added a test for aliases, and it passes with no code changes:

https://github.com/flix/flix/blob/332d8bd6c8fe700b0a0230f9f398df73b0c64ea6/main/test/flix/Test.Exp.Jvm.InvokeMethod.flix#L28

Which worries me somewhat. Have I screwed the test up somehow? If not, how is it passing without aliases being erased?

Would welcome thoughts.

Reply:
I see that. @mlutze Any idea?

Reply:
I think I've worked it out: it's my fault. The match on `Type.Cst(_, _)` doesn't match a type alias, so it's assumed to succeed.

It should therefore be fixed when I make the change that @magnus-madsen recommended to the matching.